app/services/structured_fact_importer.rb

```diff
@@ -1,7 +1,37 @@
 class StructuredFactImporter < FactImporter
+  include ActionView::Helpers::NumberHelper
+
+  MAXIMUM_FLAT_FACTS = /^(blockdevice_|ipaddress6?_|macaddress_|mtu_|speed_|auto_negotiation_|duplex_|link_|wol_)/
+
   def normalize(facts)
+    max = Setting[:maximum_structured_facts]
+    flat_counts = {}
+    dropped = 0
     # Remove empty values first, so nil facts added by flatten_composite imply compose
-    facts.delete_if { |k, v| v.nil? || (v.is_a?(String) && v.empty?) }
+    # and count flat counts for later removal and also directly remove too large sub-hashes
+    facts.delete_if do |k, v|
+      match = k.to_s.match(MAXIMUM_FLAT_FACTS)
+      if match
+        flat_counts[match.captures.first] ||= 0
+        flat_counts[match.captures.first] += 1
+      end
+      v.nil? || (v.is_a?(String) && v.empty?) || ((v.is_a?(Hash) || v.is_a?(Array)) && v.count > max && (dropped += v.count))
```

Comment:
If I understand correctly the last condition, it will remove all nodes with more than 100 children, even if none of them matches the pattern.
I think it's not what we intended.

Reply:
This is intented, see the tests. I want to drop all subtrees like ::xxx::sda (where xxx can be anything I don't care) and also mountpoint_sda (those which matches the regexp) if and only if they have more elements than 100.